src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -22,44 +23,58 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
         /// <summary>
-        /// Returns the semantic tokens data for a given document with an optional range.
+        /// Returns the semantic tokens data for a given document with an optional ranges.
         /// </summary>
         public static async Task<int[]> ComputeSemanticTokensDataAsync(
             ClientCapabilities capabilities,
             Document document,
-            LSP.Range? range,
+            LSP.Range[]? ranges,
             ClassificationOptions options,
             CancellationToken cancellationToken)
         {
-            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
+            if (ranges is not null)
+            {
+                // Expecting sorted ranges within the document
+                Array.Sort(ranges, static (r1, r2) => r1.CompareTo(r2));
+            }
 
+            var tokenTypesToIndex = SemanticTokensSchema.GetSchema(capabilities.HasVisualStudioLspCapability()).TokenTypeToIndex;
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             var text = await document.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+            using var _ = Classifier.GetPooledList(out var classifiedSpans);
 
-            // By default we calculate the tokens for the full document span, although the user 
-            // can pass in a range if they wish.
-            var textSpan = range is null ? root.FullSpan : ProtocolConversions.RangeToTextSpan(range, text);
+            // We either calculate the tokens for the full document span, or the user 
+            // can pass in a range from the full document if they wish.
+            ranges ??= new[] { ProtocolConversions.TextSpanToRange(root.FullSpan, text) };
 
-            var classifiedSpans = await GetClassifiedSpansForDocumentAsync(
-                document, textSpan, options, cancellationToken).ConfigureAwait(false);
+            foreach (var range in ranges)
+            {
+                var textSpan = ProtocolConversions.RangeToTextSpan(range, text);
+
+                var spansForRange = await GetClassifiedSpansForDocumentAsync(
+                    classifiedSpans, document, textSpan, options, cancellationToken).ConfigureAwait(false);
+            }
+
+            // Classified spans are not guaranteed to be returned in a certain order so we sort them to be safe.
+            classifiedSpans.Sort(ClassifiedSpanComparer.Instance);
 
             // Multi-line tokens are not supported by VS (tracked by https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1265495).
             // Roslyn's classifier however can return multi-line classified spans, so we must break these up into single-line spans.
-            var updatedClassifiedSpans = ConvertMultiLineToSingleLineSpans(text, classifiedSpans);
+            var updatedClassifiedSpans = ConvertMultiLineToSingleLineSpans(text, classifiedSpans.ToArray());
```

Comment:
Might be nice if you don't do the ToArray here. I think the code you are calling can just take in the SegmentedList directly, avoiding an allocation of a potentially large array.